### PR TITLE
Re-implements RandomCrop to support different crop_type

### PR DESCRIPTION
**Pull request by v-qjqs**
This PR re-implements RandomCrop to support more crop_type following the official Detectron2.
The original implementation of RandomCrop is a special case when crop_type='absolute'.

### CODE REVIEW

File: mmdet/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -581,10 +595,21 @@ class RandomCrop(object):
...
-        assert crop_size[0] > 0 and crop_size[1] > 0
+    def __init__(self,
+                 crop_size,
+                 allow_negative_crop=False,
+                 crop_type='absolute'):
+        assert crop_type in [
+            'relative_range', 'relative', 'absolute', 'absolute_range'
+        ], f'Invalid crop_type {crop_type}.'
+        if crop_type in ['absolute', 'absolute_range']:
+            assert crop_size[0] > 0 and crop_size[1] > 0
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/transforms.py:606]: we should also ensure that they are integers?

**Comment by ZwwWayne** [comment]
Please merge the newest DETR branch and resolve the comments.

File: tests/test_data/test_transform.py [hunk 2]
```diff
@@ -227,6 +227,59 @@ def area(bboxes):
     assert (area(results['gt_bboxes']) <= area(gt_bboxes)).all()
     assert (area(results['gt_bboxes_ignore']) <= area(gt_bboxes_ignore)).all()
 
+    # test assertion for invalid crop_type
+    with pytest.raises(AssertionError):
+        transform = dict(
+            type='RandomCrop', crop_type='unknow', crop_size=(1, 1))
```
Comment (ZwwWayne) [inline tests/test_data/test_transform.py:233]: unknow -> unknown

File: tests/test_data/test_transform.py [hunk 3]
```diff
@@ -227,6 +227,59 @@ def area(bboxes):
...
+        # labels
+        results['gt_labels'] = np.array([1], dtype=np.int64)
+        return results
+
+    # test crop_type "relative_range"
+    results = _construct_toy_data()
+    transform = dict(
+        type='RandomCrop', crop_type='relative_range', crop_size=(0.3, 0.7))
+    transform_module = build_from_cfg(transform, PIPELINES)
+    transform_module(copy.deepcopy(results))
```
Comment (ZwwWayne) [inline tests/test_data/test_transform.py:264]: We may also assert the size of the image, e.g., the image should be in xxx range or should have some exact value.

File: mmdet/datasets/pipelines/transforms.py [hunk 4]
```diff
@@ -581,10 +595,23 @@ class RandomCrop(object):
           `allow_negative_crop` is set to False, skip this image.
     """
 
-    def __init__(self, crop_size, allow_negative_crop=False):
-        assert crop_size[0] > 0 and crop_size[1] > 0
+    def __init__(self,
+                 crop_size,
+                 allow_negative_crop=False,
+                 crop_type='absolute'):
```
Comment (hellock) [inline mmdet/datasets/pipelines/transforms.py:601]: We may change the order of `allow_negative_crop` and `crop_type`.

File: mmdet/datasets/pipelines/transforms.py [hunk 5]
```diff
@@ -656,8 +686,58 @@ def __call__(self, results):
...
+        """
+        h, w = image_size
+        if self.crop_type == 'relative':
+            crop_h, crop_w = self.crop_size
+            return int(h * crop_h + 0.5), int(w * crop_w + 0.5)
+        elif self.crop_type == 'relative_range':
+            crop_size = np.asarray(self.crop_size, dtype=np.float32)
+            crop_h, crop_w = crop_size + np.random.rand(2) * (1 - crop_size)
+            return int(h * crop_h + 0.5), int(w * crop_w + 0.5)
+        elif self.crop_type == 'absolute':
```
Comment (hellock) [inline mmdet/datasets/pipelines/transforms.py:707]: Put the most frequently used option to the beginning:
- absolute
- absolute_range
- relative
- relative_range

**Review by hellock — APPROVED** [review]

**Pull request merged**